### PR TITLE
Asset, AssetMapping, AssetProperties and relevant entities

### PR DESCRIPTION
**Pull request by agzam**
- [X] Asset
- [x] AssetMapping https://github.com/threatgrid/ctia/pull/929
- [x] AssetProperties https://github.com/threatgrid/ctia/pull/937
- [ ] ~TargetRecord~ https://github.com/threatgrid/ctia/pull/939
       _After a discussion, it was decided to pull out Target Records related work into its own PR (not tied to Assets)_
- [ ] ~Bundle related changes~
       ~Looks like changes in CTIM schema are sufficient https://github.com/threatgrid/ctim/pull/297, and we don't have to do anything here.~
       upd: After a discussion it was decided to completely separate Bundles and Asset related entities, see: https://github.com/threatgrid/ctim/pull/303

<a name="qa">[§](#qa)</a> QA
============================
Asset, Asset Mapping, Asset Properties and ~Target Record~ contexts should be visible in Swagger UI.

Basic CRUD operations should work.

_This is preliminary work on the Assets and related entities, things are subject to change and may potentially break, once we start touching other pieces in the ecosystem, we will review and hopefully come up with proper QA strategies to conform the validity of data-flow._

---

This is a long living branch and a PR that we will be using as a temporary "master" for all Asset related work. Once every Asset feature branch/PR is merged into this and we are ready to deploy it, we will be merging this branch onto "master".

_Alternatively, we may consider merging small portions of this branch onto "master", but we will not delete this branch until all the relevant Asset pieces are done._

_This branch should always be rebased on current master, to minimize merge conflicts._

-- 

Merge and Deployment order

- [x] First merge all Asset related work in CTIM https://github.com/threatgrid/ctim/pull/294
- [x] Deploy new version of CTIM `[threatgrid/ctim "1.0.19"]`
- [x] Change CTIA dependency - threatgrid/ctim to match the deployed version
- [x] Update threatgrid/tenzin-config, merge the changes https://github.com/threatgrid/tenzin-config/pull/283
- [ ] Merge this PR
- [ ] Deploy new version of CTIA

### CODE REVIEW

File: src/ctia/entity/asset.clj [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+(def asset-mapping
+  {"asset"
+   {:dynamic false
+    :properties
+    (merge
+     em/base-entity-mapping
+     em/describable-entity-mapping
+     em/sourcable-entity-mapping
+     em/stored-entity-mapping
+     {:identity   em/tg-identity
```
Comment (ereteog) [inline src/ctia/entity/asset.clj:46]: the identity property is not in the [CTIM schema](https://github.com/threatgrid/ctim/blob/master/src/ctim/schemas/asset.cljc#L28).

**Comment by ereteog** [comment]
It could be done un a separate PR, but at some point we will need these new entities to be available in Graphql enpoint:
- define your entity graphql schema (ex for sighting: https://github.com/threatgrid/ctia/blob/master/src/ctia/entity/sighting/graphql_schemas.clj)
- add a new query type to graphql engine: (example for sighting: https://github.com/threatgrid/ctia/blob/master/src/ctia/graphql/schemas.clj#L125)
- test it in a dedicated namespace (ex for investigation: https://github.com/threatgrid/ctia/blob/master/test/ctia/http/routes/graphql/investigation_test.clj, https://github.com/threatgrid/ctia/blob/master/test/data/investigation.graphql)

File: src/ctia/entity/asset_mapping.clj [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+(def asset-mapping-mapping
+  {"asset-mapping"
+   {:dynamic false
+    :properties
+    (merge
+     em/base-entity-mapping
+     em/sourcable-entity-mapping
+     em/stored-entity-mapping
+     {:identity    em/tg-identity
```
[the remark this answers is not in the document]
Comment (agzam) [reply src/ctia/entity/asset_mapping.clj]: Good catch. This shouldn't be here. The result of copy-pasta. I think I initially borrowed it from Actor entity.

**Comment by jyoverma** [comment]
Ag, could you add asset_ref as a searchable field in asset_mapping and asset_properties?

**Comment by ereteog** [comment]
> Ag, could you add asset_ref as a searchable field in asset_mapping and asset_properties?

@jyoverma what do you mean by searchable?

The mapping looks fine. The detail is that you must specify the field that you query like: `assert_ref:"http://ex.tld/ctia/asset/asset-61884b14-e273-4930-a5ff-dcce69207724"` (the `"` are important, unless you would have to escape reserved lucene characters in the value like: `:`).
We avoid to use `all_token` for fields that contain ids because it includes it in what Elasticsearch calls the [`_all` field](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/mapping-all-field.html). This `_all` field is indexed as a `text` whatever is in it and thus `http://ex.tld/ctia/asset/asset-61884b14-e273-4930-a5ff-dcce69207724` will be splitted in the following tokens: `http`, `ex`, `tld`, `ctia`, `asset`, `asset`, `61884b14`, `e273`, `4930`, `a5ff`, `dcce69207724`. Then each of these token could be matched individually in any order with a `classic` query like typical users use, such as the following queries: `http`, `asset ctia`, `ex` which is meaningless.

**Comment by ereteog** [comment]
> Ag, could you add asset_ref as a searchable field in asset_mapping and asset_properties?

Jyoti, I deleted my previous remark, I tested the branch and understood your purpose. Ag you just need to add `asset_ref` in each SearchParams schema in order to enable filtering on this field in the search route.

File: project.clj [hunk 3]
```diff
@@ -48,7 +48,8 @@
                  [metosin/schema-tools "0.12.2"]
                  [threatgrid/flanders "0.1.23"]
 
-                 [threatgrid/ctim "1.0.17"]
+                 ;; TODO: Fix the version before merging
+                 [threatgrid/ctim "1.0.19-SNAPSHOT"]
```
Comment (ereteog) [inline project.clj:52]: you need to release CTIM 1.0.19

**Review by ereteog — COMMENTED** [review]
You are nearly good to go:
- I proposed minor changes.
- You need to publish CTIM 1.0.19 and before publishing I did a small PR about the model, we should review it with @jyoverma.
- you need to make a PR on tenzin-config to update the store properties on each environment / zone / instance type: https://github.com/threatgrid/tenzin-config/tree/master/ctia, https://github.com/threatgrid/tenzin-config/tree/master/private-ctia, https://github.com/threatgrid/tenzin-config/tree/master/staging-ctia

File: src/ctia/entity/asset_properties.clj [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+(s/defschema AssetPropertiesFieldsParam
+  {(s/optional-key :fields) [asset-properties-sort-fields]})
+
+(s/defschema AssetPropertiesByExternalIdQueryParams
+  (st/merge
+   routes.common/PagingParams
+   AssetPropertiesFieldsParam))
+
+(def asset-properties-enumerable-fields
+  [])
```
[the remark this answers is not in the document]
Comment (agzam) [reply src/ctia/entity/asset_properties.clj]: addressed in ca1634daab261acac4a0085768cba9feea4a723e

File: test/ctia/entity/asset_mapping_test.clj [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                                                         :headers {"Authorization" "45c1f5e3f05d0"})]
+                                           (is (= 200 (:status response)))
+                                           (is (= expected (check-fn response)) desc))
+
+        "specificity:\"unique\""
+        (fn [r] (-> r :parsed-body first :specificity))
+        (-> asset-mapping-sample :specificity)
+        "Searching Asset Mapping by specificity"
+
+        ;; TODO: Add more cases
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/entity/asset_mapping_test.clj]: addressed in ca1634daab261acac4a0085768cba9feea4a723e

File: test/ctia/entity/asset_properties_test.clj [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+        "*"
+        (fn [r] (-> r :parsed-body first :properties first :name))
+        (-> asset-properties-sample :properties first :name)
+        "Properties name match"
+
+        "*"
+        (fn [r] (-> r :parsed-body first :properties first :value))
+        (-> asset-properties-sample :properties first :value)
+        "Properties value match"
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/entity/asset_properties_test.clj:43]: addressed in ca1634daab261acac4a0085768cba9feea4a723e

File: test/ctia/entity/asset_test.clj [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                                                                     :headers {"Authorization" "45c1f5e3f05d0"})]
+                                           (is (= 200 (:status response)))
+                                           (is (= expected (check-fn response)) desc))
+
+        "asset_type:\"device\""
+        (fn [r] (-> r :parsed-body first :asset_type))
+        (-> asset-sample :asset_type)
+        "Searching by an Asset type works"
+
+        ;; TODO: Add more cases
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/entity/asset_test.clj]: addressed in ca1634daab261acac4a0085768cba9feea4a723e

**Pull request merged**